### PR TITLE
CORE-18342 custom ordering for named vault queries

### PR DESCRIPTION
**Pull request by blsemo**
Initial version of NamedVaultQueries with an order by clause. So far, they work in principle but need a lot more test coverage.

### CODE REVIEW

File: .run/Combined Worker Local.run.xml [hunk 1]
```diff
@@ -1,7 +1,7 @@
 <component name="ProjectRunConfigurationManager">
   <configuration default="false" name="Combined Worker Local" type="JarApplication">
     <option name="JAR_PATH" value="$PROJECT_DIR$/applications/workers/release/combined-worker/build/bin/corda-combined-worker-5.2.0.0-SNAPSHOT.jar" />
-    <option name="VM_PARAMETERS" value="-Dco.paralleluniverse.fibers.verifyInstrumentation=true" />
+    <option name="VM_PARAMETERS" value="-agentlib:jdwp=transport=dt_socket,server=y,suspend=n,address=*:5005 -Dco.paralleluniverse.fibers.verifyInstrumentation=true" />
```
Comment (rick-r3) [inline .run/Combined Worker Local.run.xml]: Assume this is unrelated?

File: components/ledger/ledger-persistence/src/main/kotlin/net/corda/ledger/persistence/query/registration/impl/VaultNamedQueryBuilderImpl.kt [hunk 2]
```diff
@@ -45,37 +63,68 @@ class VaultNamedQueryBuilderImpl(
...
+                parseOrderBy()
             )
         )
     }
+
+    private fun prepareQuery(): VaultNamedQuery.ParsedQuery {
+        val notNullQuery =
+            "${requireNotNull(query) { "Vault named query: $name does not have its query statement set" }}${
+                if (unconumedStatesOnly) {
+                    " AND (visible_states.consumed IS NULL OR visible_states.consumed < :${TIMESTAMP_LIMIT_PARAM_NAME})"
```
Comment (rick-r3) [inline components/ledger/ledger-persistence/src/main/kotlin/net/corda/ledger/persistence/query/registration/impl/VaultNamedQueryBuilderImpl.kt]: I think this should be `OR visible_states.consumed >= :${TIMESTAMP_LIMIT_PARAM_NAME})`.  i.e. greater than (or equal to).  You're not interested if they were consumed before the `created` cut off time.

File: components/flow/flow-service/src/main/kotlin/net/corda/flow/application/persistence/query/StableResultSetImpl.kt [hunk 3]
```diff
@@ -47,8 +49,18 @@ data class StableResultSetImpl<R> internal constructor(
...
 
-        resumePoint = nextResumePoint
+        check( nextResumePoint == null || numberOfRowsFromQuery == null ){
+            "Invalid query handler - can either have offset or resume point, but both were set."
+        }
+
 
+        resumePoint = nextResumePoint
+        if (numberOfRowsFromQuery != null) {
+            hasNext = numberOfRowsFromQuery != 0  // If there are no rows left there are no more pages to fetch
```
Comment (rick-r3) [inline components/flow/flow-service/src/main/kotlin/net/corda/flow/application/persistence/query/StableResultSetImpl.kt:59]: I think we abandon use of `hasNext` and actually continue to use presence or otherwise of `nextResumePoint`.  I believe this existing logic will suggest there is an extra page when the `numRows MOD limit == 0`.  e.g. with 100 rows and page size 100, it will think there are 2 pages, one with 100 rows and one with zero rows.

File: components/flow/flow-service/src/main/kotlin/net/corda/flow/application/persistence/query/StableResultSetImpl.kt [hunk 4]
```diff
@@ -47,8 +49,18 @@ data class StableResultSetImpl<R> internal constructor(
             "Infinite query detected; resume point has not been updated"
         }
 
-        resumePoint = nextResumePoint
+        check( nextResumePoint == null || numberOfRowsFromQuery == null ){
```
Comment (rick-r3) [inline components/flow/flow-service/src/main/kotlin/net/corda/flow/application/persistence/query/StableResultSetImpl.kt:52]: When allowing `nextResumePoint` even for custom ordered queries as suggested below, need to remove this `check`.

File: components/ledger/ledger-persistence/src/main/kotlin/net/corda/ledger/persistence/query/execution/impl/VaultNamedQueryExecutorImpl.kt [hunk 5]
```diff
@@ -226,7 +250,7 @@ class VaultNamedQueryExecutorImpl(
                 with(rawResults) {
                     return ProcessedQueryResults(
                         results.map { it.stateAndRef },
-                        if (hasMore) results.last().resumePoint else null,
+                        if (hasMore && !useOffset) results.last().resumePoint else null,
```
Comment (rick-r3) [inline components/ledger/ledger-persistence/src/main/kotlin/net/corda/ledger/persistence/query/execution/impl/VaultNamedQueryExecutorImpl.kt:253]: Recommend removing `useOffset` check here and always set the resume point if `hasMore`.

File: components/ledger/ledger-persistence/src/main/kotlin/net/corda/ledger/persistence/query/execution/impl/VaultNamedQueryExecutorImpl.kt [hunk 6]
```diff
@@ -251,7 +275,7 @@ class VaultNamedQueryExecutorImpl(
 
                     return ProcessedQueryResults(
                         filteredRawData.map { it.stateAndRef },
-                        if (moreResults) filteredRawData.last().resumePoint else null,
+                        if (moreResults && !useOffset) filteredRawData.last().resumePoint else null,
```
Comment (rick-r3) [inline components/ledger/ledger-persistence/src/main/kotlin/net/corda/ledger/persistence/query/execution/impl/VaultNamedQueryExecutorImpl.kt:278]: Recommend removing `useOffset` check here and always set the resume point if `moreResults`.

File: components/ledger/ledger-persistence/src/main/kotlin/net/corda/ledger/persistence/query/registration/impl/VaultNamedQueryBuilderImpl.kt (hunk 2 above)
Comment (dickon) [reply components/ledger/ledger-persistence/src/main/kotlin/net/corda/ledger/persistence/query/registration/impl/VaultNamedQueryBuilderImpl.kt]: fixed in my version of this PR

File: .run/Combined Worker Local.run.xml (hunk 1 above)
Comment (dickon) [reply .run/Combined Worker Local.run.xml]: reverted in my version of this PR (https://github.com/corda/corda-runtime-os/pull/5691/files)
